File: internal/domain/i18n/duration.go
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

package i18n

import (
	"fmt"
	"strings"
	"time"
)

var humanizeUnits = []struct {
	minutes int64
	key     string
}{
	{365 * 24 * 60, "time.year"},
	{30 * 24 * 60, "time.month"},
	{24 * 60, "time.day"},
	{60, "time.hour"},
	{1, "time.minute"},
}

const humanizeParts = 2

func HumanizeDuration(locale string, d time.Duration) string {
	minutes := int64(d / time.Minute)
	if minutes < 1 {
		return T(locale, "time.less_than_minute")
	}
	parts := make([]string, 0, humanizeParts)
	for _, unit := range humanizeUnits {
		n := minutes / unit.minutes
		if n == 0 {
			continue
		}
		parts = append(parts, humanizePart(locale, n, unit.key))
		minutes %= unit.minutes
		if len(parts) == humanizeParts {
			break
		}
	}
	return strings.Join(parts, ", ")
}

func humanizePart(locale string, n int64, key string) string {
	if n != 1 {
		key += "s"
	}
	return fmt.Sprintf("%d %s", n, T(locale, key))
}

File: web/admin/src/lib/duration.ts
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

const UNITS = ['ns', 'µs', 'ms', 's'] as const;

export function durationLabel(ns: number): string {
  let value = Math.max(0, ns);
  let unit = 0;
  while (value >= 1000 && unit < UNITS.length - 1) {
    value /= 1000;
    unit++;
  }
  return `${value.toFixed(2)} ${UNITS[unit]}`;
}
